File: Backend API/src/services/llmService.js
const { GoogleGenAI } = require('@google/genai');

// Initialize Gemini client conditionally to not crash if key is missing initially.
let ai = null;

function initializeAI() {
    if (!ai && process.env.GEMINI_API_KEY) {
        ai = new GoogleGenAI({ apiKey: process.env.GEMINI_API_KEY });
    }
    return ai;
}

/**
 * Generate improvement suggestions for a resume based on the Job Description and calculated scores.
 * @param {string} resumeText - The parsed text of the candidate's resume
 * @param {string} jobDescription - The job description
 * @param {number} atsScore - The calculated ATS score
 * @param {Array<string>} missingSkills - The missing skills identified by the ML engine
 * @returns {Promise<string>}
 */
async function generateImprovementSuggestions(resumeText, jobDescription, atsScore, missingSkills) {
    const client = initializeAI();
    if (!client) {
        return "💡 API Key missing. Please add the GEMINI_API_KEY in the .env file to enable AI suggestions.";
    }

    const hasJD = !!jobDescription && jobDescription.trim() !== "" && jobDescription.trim() !== "{}";
    const prompt = hasJD ? `
You are an expert technical recruiter and ATS system optimizer.
Please provide 3-4 specific, actionable bullet points to improve the following resume for the provided Job Description.

Target Job Description:
${jobDescription}

Candidate Resume:
${extrapolateResume(resumeText)}

Current ATS Score: ${atsScore}/100
Missing Skills (Identified by basic semantic check): ${missingSkills && missingSkills.length ? missingSkills.join(", ") : "None"}

Keep your response concise and formatted in markdown bullet points. Do not include introductory text like "Here are some suggestions:" - just provide the bullet points directly.
` : `
You are an expert technical recruiter and resume reviewer.
Please provide 3-4 specific, actionable bullet points to generally improve the formatting, impact, and ATS-readability of the following resume.

Candidate Resume:
${extrapolateResume(resumeText)}

Current General Processing Score: ${atsScore}/100

Keep your response concise and formatted in markdown bullet points. Do not include introductory text like "Here are some suggestions:" - just provide the bullet points directly. Focus on action verbs, quantifiable achievements, and clear formatting.
`;

    try {
        const response = await client.models.generateContent({
            model: 'gemini-2.5-flash',
            contents: prompt,
        });
        return response.text;
    } catch (error) {
        console.error("Gemini Suggestion Error:", error);
        return "💡 Error generating insights from the AI model. Please try again later.";
    }
}

/**
 * Handles conversational queries about the resume and job description.
 * @param {string} resumeText
 * @param {string} jobDescription
 * @param {Array<{role: string, text: string}>} chatHistory - Previous chat messages
 * @param {string} userMessage - The new question from the user
 */
async function handleChatQuery(resumeText, jobDescription, chatHistory, userMessage) {
    const client = initializeAI();
    if (!client) {
        return "💡 API Key missing. Please add the GEMINI_API_KEY in the .env file to enable AI chat.";
    }

    const hasJD = !!jobDescription && jobDescription.trim() !== "" && jobDescription.trim() !== "{}";
    const systemInstruction = hasJD ? `
You are an expert technical recruiter and AI assistant for a Resume Tracking application.
Your goal is to help candidates improve their resume and understand the job description they are applying for.

Here is the context for this conversation:
Job Description:
${jobDescription}

Candidate Resume:
${extrapolateResume(resumeText)}

Please answer the user's question concisely, referencing the context above when appropriate.
` : `
You are an expert technical recruiter and AI assistant for a Resume Tracking application.
Your goal is to help candidates improve the formatting, readability, and impact of their resume generally.

Here is the context for this conversation:
Candidate Resume:
${extrapolateResume(resumeText)}

Please answer the user's question concisely, referencing the context above when appropriate. Focus your advice on general resume best practices.
`;

    try {
        // Format the chat history for Gemini
        const contents = chatHistory.map(msg => ({
            role: msg.role === 'user' ? 'user' : 'model',
            parts: [{ text: msg.text }]
        }));

        // Add the new user message
        contents.push({
            role: 'user',
            parts: [{ text: systemInstruction + "\n\nUser Question: " + userMessage }]
        });

        const response = await client.models.generateContent({
            model: 'gemini-2.5-flash',
            contents: contents,
        });

        return response.text;
    } catch (error) {
        console.error("Gemini Chat Error:", error);
        return "💡 Error generating a response from the AI model. Please try again later.";
    }
}

function extrapolateResume(text) {
    if (!text) return "No resume text provided.";
    return text.length > 2000 ? text.substring(0, 2000) + "..." : text;
}

module.exports = {
    generateImprovementSuggestions,
    handleChatQuery
};

File: ML/llm_service.py
import os
from google import genai
from dotenv import load_dotenv

# Load environment variables
backend_env_path = os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))), 'Backend API', '.env')
load_dotenv(backend_env_path)

GEMINI_API_KEY = os.getenv("GEMINI_API_KEY")

if not GEMINI_API_KEY:
    print("WARNING: GEMINI_API_KEY not found in .env")

# Use gemini-2.0-flash
MODEL_NAME = 'gemini-2.0-flash'

def get_llm_client():
    if not GEMINI_API_KEY:
        return None
    return genai.Client(api_key=GEMINI_API_KEY)

def generate_resume_suggestions(resume_text: str, jd_text: str) -> str:
    """
    Analyzes the resume against the job description and provides actionable suggestions.
    """
    if not GEMINI_API_KEY:
        return "LLM API Key not configured."

    client = get_llm_client()
    if not client:
        return "LLM Client failed to initialize."
        
    prompt = f"""
You are an expert technical recruiter and ATS optimization specialist.
Analyze the following resume against the provided Job Description.

Job Description:
{jd_text}

Resume:
{resume_text}

Provide concise, actionable suggestions on how the candidate can improve their resume to better match the job description. 
Focus on:
1. Missing keywords or skills to add (if they possess them)
2. Impact formatting (e.g., adding metrics to bullet points)
3. Structural improvements
Keep your response short, formatted in markdown, and highly professional. Do not write a cover letter or rewrite the resume for them.
"""
    
    try:
        response = client.models.generate_content(
            model=MODEL_NAME,
            contents=prompt
        )
        return response.text
    except Exception as e:
        print(f"Error generating suggestions: {e}")
        return f"Error connecting to LLM service: {str(e)}"

def answer_technical_question(resume_text: str, question: str) -> str:
    """
    Answers a specific user question about the candidate's resume based on the extracted text.
    """
    if not GEMINI_API_KEY:
        return "LLM API Key not configured."

    client = get_llm_client()
    if not client:
        return "LLM Client failed to initialize."
        
    prompt = f"""
You are an AI assistant helping a recruiter review a resume.
Answer the following question based ONLY on the provided resume text. If the resume does not contain the answer, say "The resume does not specify this information."

Resume Text:
{resume_text}

Question:
{question}
"""
    
    try:
        response = client.models.generate_content(
            model=MODEL_NAME,
            contents=prompt
        )
        return response.text
    except Exception as e:
        print(f"Error answering question: {e}")
        return f"Error connecting to LLM service: {str(e)}"
